module/zfs/dnode.c

```diff
@@ -1421,13 +1440,14 @@ dnode_setdirty(dnode_t *dn, dmu_tx_t *tx)
 	 */
 	dmu_objset_userquota_get_ids(dn, B_TRUE, tx);
 
-	mutex_enter(&os->os_lock);
+	dnl = os->os_dnode_list + (CPU_SEQID % OS_DNODE_LISTS);
```

Comment:
If I understand correctly the big win from this patch is from breaking up the `os->os_lock` lock to reduce contention.  A nice way to do that would be to use the `dn->dn_object` to select the sub-list rather than the CPU_SEQID to generate an index.

This way consecutively allocated objects which are all immediately dirtied should have minimal contention.  Plus it means given the `dnode_t` latter in `dnode_free()` you can easily lock the correct sub-list which will resolve the failure observed during automated testing.